test/unary_handler_test.go

```diff
@@ -0,0 +1,170 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := p2.Connect(peer1ID, peer1Addrs); err != nil {
+		t.Fatal(err)
+	}
+
+	var proto protocol.ID = "sqrt"
+	if err := p1.AddUnaryHandler(
+		proto,
+		func(ctx context.Context, data []byte) ([]byte, error) {
+			f := float64FromBytes(data)
+			if f < 0 {
+				return nil, fmt.Errorf("can't extract square root from negative")
+			}
+
+			result := math.Sqrt(f)
+			return float64Bytes(result), nil
+		},
+	); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run(
+		"test bad request",
+		func(t *testing.T) {
+			_, err := p2.CallUnaryHandler(context.Background(), peer1ID, proto, float64Bytes(-64))
+			if err == nil {
+				t.Fatal("remote should have returned error")
+			}
+			t.Logf("remote correctly returned error: '%v'\n", err)
+		},
+	)
+
+	t.Run(
+		"test correct request",
+		func(t *testing.T) {
+			reply, err := p2.CallUnaryHandler(context.Background(), peer1ID, proto, float64Bytes(64))
+			if err != nil {
+				t.Fatal(err)
+			}
+			result := float64FromBytes(reply)
+			t.Logf("remote returned: %f\n", result)
+		},
+	)
+
+	t.Run(
+		"test bad proto",
+		func(t *testing.T) {
+			_, err := p2.CallUnaryHandler(context.Background(), peer1ID, "bad proto", make([]byte, 0))
+			if err == nil {
+				t.Fatal("expected error")
+			}
+			t.Logf("remote correctly returned error: '%v'\n", err)
+		},
+	)
+}
+
+func TestCancellation(t *testing.T) {
+	_, p1, cancel1 := createDaemonClientPair(t)
+	_, p2, cancel2 := createDaemonClientPair(t)
+
+	t.Cleanup(func() {
+		cancel1()
+		cancel2()
+	})
+
+	peer1ID, peer1Addrs, err := p1.Identify()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := p2.Connect(peer1ID, peer1Addrs); err != nil {
+		t.Fatal(err)
+	}
+
+	var proto protocol.ID = "sqrt"
+	if err := p1.AddUnaryHandler(proto, sqrtHandler); err != nil {
+		t.Fatal(err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	_, err = p2.CallUnaryHandler(ctx, peer1ID, proto, []byte("hi"))
+	if err == nil {
+		t.Fatal("expected error")
+	}
+}
+
+func TestAddUnaryHandler(t *testing.T) {
+	// create a singe daemon and connect two client to it
+	dmaddr, c1maddr, dir1Closer := getEndpointsMaker(t)(t)
+	_, c2maddr, dir2Closer := getEndpointsMaker(t)(t)
+
+	daemon, closeDaemon := createDaemon(t, dmaddr)
+
+	c1, closeClient1 := createClient(t, daemon.Listener().Multiaddr(), c1maddr)
+	c2, closeClient2 := createClient(t, daemon.Listener().Multiaddr(), c2maddr)
+
+	defer func() {
+		closeClient1()
+		closeClient2()
+
+		closeDaemon()
+
+		dir1Closer()
+		dir2Closer()
+	}()
+
+	var proto protocol.ID = "sqrt"
+
+	if err := c1.AddUnaryHandler(proto, sqrtHandler); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := c2.AddUnaryHandler(proto, sqrtHandler); err == nil {
+		t.Fatal("adding second unary handler with same name should have returned error")
+	}
+
+	if err := c1.Close(); err != nil {
+		t.Fatal("closing client 1 should not have returned an error")
+	}
+
+	time.Sleep(time.Second * 3)
```

Comment:
```suggestion
	time.Sleep(time.Second)
```

If possible, let's limit the delays by 1 sec to save the CI time :)